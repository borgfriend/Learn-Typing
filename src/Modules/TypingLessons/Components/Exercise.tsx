import React, { useEffect } from "react";
import { ExerciseChar } from "../../../components/ExerciseChar/ExerciseChar";
import { useDispatch } from "react-redux";
import {
  endLesson,
  startLesson,
  CharField
} from "../../../Stores/ExerciseStore";

export const Exercise: React.FC<{lesson: CharField[]}> = ({lesson =[]}) => {
  const dispatch = useDispatch();
  const [exerciseFields, setExerciseFields] = React.useState(lesson);
  const [currentIndex, setCurrentIndex] = React.useState(0);

  const isLastIndex = currentIndex === exerciseFields.length

  useEffect(() => {
    const handleKeyPress = (e: KeyboardEvent) => {
      if (e.key === "Space" && e.target === document.body) {
        e.preventDefault();
      }

      if (isLastIndex) {
        dispatch(endLesson({mistakeCount: calculateMistakes(exerciseFields)}));
        return;
      }
      const keyValue =(e.key === "Enter") ? "\n" : e.key;

      const newExerciseFields = [...exerciseFields];
      const field = exerciseFields[currentIndex]
      newExerciseFields[currentIndex] = {
        color: (field.value === keyValue) ? "ok" : "error",
        value: field.value,
      }
      setExerciseFields(newExerciseFields)
      setCurrentIndex(currentIndex+1)

      if (currentIndex === 1) {
        dispatch(startLesson());
      }
    };

    document.addEventListener("keydown", handleKeyPress);

    return () => {
      document.removeEventListener("keydown", handleKeyPress);
    };
  }, [dispatch, isLastIndex, exerciseFields, currentIndex]);

  const text = exerciseFields.map((item, key) => (
    <ExerciseChar key={key} data={item} isCurrent={key === currentIndex} />
  ));

  return <section>{text}</section>;
};

const calculateMistakes = (lesson: CharField[]) => {
  return lesson.filter((val) => val.color === "error").length;
};
