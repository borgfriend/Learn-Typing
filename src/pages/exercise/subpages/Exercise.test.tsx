import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { Exercise } from "./Exercise";

it("renders an empty exercise", () => {
  const mockFn = jest.fn();
  const { container } = render(
    <Exercise lesson={[]} completeLesson={mockFn}></Exercise>
  );
  expect(container).toMatchSnapshot();
});

it("renders an exercise", () => {
  const mockFn = jest.fn();
  const { container } = render(
    <Exercise
      lesson={[
        { value: "a", color: "transparent" },
        { value: "b", color: "transparent" },
      ]}
      completeLesson={mockFn}
    ></Exercise>
  );
  expect(container).toMatchSnapshot();

  fireEvent.keyDown(document, { key: "a", code: 56, charCode: 56 });
  fireEvent.keyDown(document.body, { key: "Space", code: 32, charCode: 32 });
  fireEvent.keyDown(document, { key: "Enter", code: 13, charCode: 13 });

  expect(container).toMatchSnapshot();
  expect(mockFn).toHaveBeenCalled();
});
