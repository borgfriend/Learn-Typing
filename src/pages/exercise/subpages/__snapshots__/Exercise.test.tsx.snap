// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders an empty exercise 1`] = `
<div>
  <section />
</div>
`;

exports[`renders an exercise 1`] = `
<div>
  <section>
    <span
      class="sc-FEMpB sc-fLDLck cRvibt eWGIFq"
    >
      a
    </span>
    <span
      class="sc-FEMpB cRvibt"
    >
      b
    </span>
  </section>
</div>
`;

exports[`renders an exercise 2`] = `
<div>
  <section>
    <span
      class="sc-FEMpB sc-gjcoXW cRvibt bdPGgx"
    >
      a
    </span>
    <span
      class="sc-FEMpB sc-dVBluf cRvibt bNzGEa"
    >
      b
    </span>
  </section>
</div>
`;
